### PR TITLE
[FLINK-14881] [s3|kinesis] Add support for IAM Roles for Service Accounts on AWS EKS

### PR DESCRIPTION
**Pull request by aroch**
## What is the purpose of the change

IAM Roles for Service Accounts have many advantages when deploying Flink on AWS EKS.
From AWS documentation:
> With IAM roles for service accounts on Amazon EKS clusters, you can associate an IAM role with a Kubernetes service account. This service account can then provide AWS permissions to the containers in any pod that uses that service account. With this feature, you no longer need to provide extended permissions to the worker node IAM role so that pods on that node can call AWS APIs.
 
In order to support for IAM Roles for Service Accounts we need to add support for a relatively new authentication method in AWS SDK called WebIdentityToken.

This pull request adds support for WebIdentityToken authentication method when using AWS services. This includes S3 and Kinesis.

## Brief change log

S3
  - Bumping AWS SDK dependency to 1.11.754
  - Bumping httpcomponents:httpclient dependency to 4.5.9
  - Added aws-java-sdk-sts dependency

Kinesis
  - Bumping AWS SDK dependency to 1.11.754
  - Added CredentialProvider type for WebIdentityToken

## Verifying this change

It is tricky to verify this change by end-to-end test as Kinesalite doesn't support authentication methods. Minio does support it but requires extra containers to be launched to do the actual token verification that will complicate and lengthen the s3 tests. 

Added configuration unit tests in `AWSUtilTest`

As this change is mainly a dependency change, I verified it manually on an actual EKS cluster.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**yes** / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no**)
  - The runtime per-record code paths (performance sensitive): (yes / **no**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (**yes** / no)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

Currently I could not find the deployment on EKS documented anywhere. If needed i'm willing to add docs if that would be helpful.

### CODE REVIEW

**Comment by aroch** [comment]
@flinkbot run azure

**Comment by rmetzger** [comment]
Thanks a lot for your contribution. I manually re-triggered the failed job on Azure: https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=721&view=results

**Comment by rmetzger** [comment]
CI has passed ✅

**Comment by aroch** [comment]
Thanks @rmetzger!

**Comment by rmetzger** [comment]
@tweise (kinesis) @pnowojski (s3): What's your take on this change? 
I'm tempted to merge it without additional testing. @aroch has tested the change, and it is unlikely to break anything.

**Comment by aroch** [comment]
It would be great if someone could take a second look.
I verified on EKS with the Lyft k8s operator.

**Comment by rmetzger** [comment]
Okay, looks like nobody has time for this. I will test it myself, because it would be a bummer if this change didn't make it into 1.11.0

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/AWSConfigConstants.java [hunk 1]
```diff
@@ -70,15 +73,18 @@
 	/** Optional configuration for profile name if credential provider type is set to be PROFILE. */
 	public static final String AWS_PROFILE_NAME = profileName(AWS_CREDENTIALS_PROVIDER);
 
-	/** The role ARN to use when credential provider type is set to ASSUME_ROLE. */
+	/** The role ARN to use when credential provider type is set to ASSUME_ROLE or WEB_IDENTITY_TOKEN. */
 	public static final String AWS_ROLE_ARN = roleArn(AWS_CREDENTIALS_PROVIDER);
 
-	/** The role session name to use when credential provider type is set to ASSUME_ROLE. */
+	/** The role session name to use when credential provider type is set to ASSUME_ROLE or WEB_IDENTITY_TOKEN. */
```
Comment (rmetzger) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/AWSConfigConstants.java:79]: Can you document this also in `docs/dev/connectors/kinesis.md`?

Comment (rmetzger) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/AWSConfigConstants.java:79]: I mean in general that this authentication method exists, not this particular line of code :)

**Comment by tweise** [comment]
@rmetzger I don't have time to test this change, but I also think it should be merged before 1.11.0 so that any hidden issues can be caught as part of the release candidate testing. As we have just discovered with the kinesis producer in 1.10, benign dependency changes can have unforeseen consequences, so putting this through a thorough benchmark would be in order before release.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/AWSConfigConstants.java (hunk 1 above)
Comment (aroch) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/AWSConfigConstants.java:79]: I saw there was no dedicated section for setting up access to Kinesis. I thought it makes sense to add a new section with some details.
I'm not a technical writer, so please let me know what you think :)

**Comment by rmetzger** [comment]
@aroch Thanks a lot for updating the docs! I'm still working on this.
I'm struggling my way through learning EKS, building custom Flink docker images etc. :) 
Right now, I'm facing `org.apache.flink.kinesis.shaded.com.amazonaws.services.kinesis.producer.DaemonException: The child process has been shutdown and can no longer accept messages.`. From my initial research, I assume that my alpine-based docker image (build from `flink-container/docker` won't work with the Kinesis binaries).
How are you building your 1.11-SNAPSHOT docker containers?

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/AWSConfigConstants.java (hunk 1 above)
Comment (rmetzger) [reply flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/AWSConfigConstants.java:79]: Thanks a lot. It reads pretty well!

**Comment by rmetzger** [comment]
Actually, I managed to get a debian based docker image. I still have the mentioned exception though.

File: flink-filesystems/flink-s3-fs-presto/src/main/resources/META-INF/NOTICE [hunk 2]
```diff
@@ -12,11 +12,11 @@ This project bundles the following dependencies under the Apache Software Licens
...
-- com.amazonaws:aws-java-sdk-core:1.11.271
-- com.amazonaws:aws-java-sdk-dynamodb:1.11.271
-- com.amazonaws:aws-java-sdk-kms:1.11.271
-- com.amazonaws:aws-java-sdk-s3:1.11.271
-- com.amazonaws:jmespath-java:1.11.271
+- com.amazonaws:aws-java-sdk-core:1.11.754
+- com.amazonaws:aws-java-sdk-dynamodb:1.11.754
+- com.amazonaws:aws-java-sdk-kms:1.11.754
+- com.amazonaws:aws-java-sdk-s3:1.11.754
+- com.amazonaws:jmespath-java:1.11.754
```
Comment (rmetzger) [inline flink-filesystems/flink-s3-fs-presto/src/main/resources/META-INF/NOTICE:19]: The `com.amazonaws:aws-java-sdk-sts:1.11.754` dependency is missing here?

File: flink-filesystems/flink-s3-fs-hadoop/src/main/resources/META-INF/NOTICE [hunk 3]
```diff
@@ -3,11 +3,11 @@ Copyright 2014-2020 The Apache Software Foundation
...
 
-- com.amazonaws:aws-java-sdk-core:1.11.271
-- com.amazonaws:aws-java-sdk-dynamodb:1.11.271
-- com.amazonaws:aws-java-sdk-kms:1.11.271
-- com.amazonaws:aws-java-sdk-s3:1.11.271
-- com.amazonaws:jmespath-java:1.11.271
+- com.amazonaws:aws-java-sdk-core:1.11.754
+- com.amazonaws:aws-java-sdk-dynamodb:1.11.754
+- com.amazonaws:aws-java-sdk-kms:1.11.754
+- com.amazonaws:aws-java-sdk-s3:1.11.754
```
Comment (rmetzger) [inline flink-filesystems/flink-s3-fs-hadoop/src/main/resources/META-INF/NOTICE:9]: The com.amazonaws:aws-java-sdk-sts:1.11.754 dependency is missing here?

**Pull request closed without merging**